pkg/dryrun/dryrun.go

```diff
@@ -0,0 +1,810 @@
+			}
+		} else {
+			if err := comp.Delete(d.cluster); err != nil {
+				logrus.Errorf("Failed to cleanup %v. %v", comp.Name(), err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (d *DryRun) simulateK8sNode() error {
+	if d.realClient == nil {
+		node := &v1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   "mockNode",
+				Labels: make(map[string]string),
+			},
+			Status: v1.NodeStatus{
+				Allocatable: map[v1.ResourceName]resource.Quantity{
+					v1.ResourcePods: resource.MustParse(strconv.Itoa(1)),
+				},
+			},
+		}
+		if err := d.mockClient.Create(context.TODO(), node); err != nil {
+			return err
+		}
+	} else {
+		nodeList := &v1.NodeList{}
+		err := d.realClient.List(context.TODO(), nodeList)
+		if err != nil || nodeList == nil || len(nodeList.Items) == 0 {
+			return fmt.Errorf("failed to get real k8s node, err %v, nodeList %+v", err, nodeList)
+		}
+
+		for _, node := range nodeList.Items {
+			d.cleanupObject(&node)
+			err = d.mockClient.Create(context.TODO(), &node)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (d *DryRun) cleanupObject(obj client.Object) error {
+	obj.SetGenerateName("")
+	obj.SetUID("")
+	obj.SetResourceVersion("")
+	obj.SetGeneration(0)
+	obj.SetSelfLink("")
+	obj.SetCreationTimestamp(metav1.Time{})
+	obj.SetFinalizers(nil)
+	obj.SetOwnerReferences(nil)
+	obj.SetClusterName("")
+	obj.SetManagedFields(nil)
+
+	return nil
+}
+
+func (d *DryRun) validateObjects(dsObjs, operatorObjs []client.Object, h *migration.Handler) error {
+	// object kind, object name to object
+	operatorObjMap := make(map[string]map[string]client.Object)
+	for _, obj := range operatorObjs {
+		kind := obj.GetObjectKind().GroupVersionKind().Kind
+		if _, ok := operatorObjMap[kind]; !ok {
+			operatorObjMap[kind] = make(map[string]client.Object)
+		}
+		operatorObjMap[kind][obj.GetName()] = obj
+	}
+
+	var lastErr error
+	for _, dsObj := range dsObjs {
+		kind := dsObj.GetObjectKind().GroupVersionKind().Kind
+		name := dsObj.GetName()
+
+		if skipObjectKinds[kind] {
+			continue
+		}
+
+		// Map daemonset-installed object to operator-installed object
+		if kind == "DaemonSet" && name == "portworx" {
+			// Portworx DaemonSet is deployed as Pod by operator
+			kind = "Pod"
+			logrus.Info("Portworx will be deployed via Pod scheduled by operator (instead of DaemonSet)")
+		} else if kind == "Deployment" && name == "px-csi-ext" {
+			if m, ok := operatorObjMap["StatefulSet"]; ok {
+				if _, ok := m[name]; ok {
+					kind = "StatefulSet"
+					logrus.Info("px-csi-ext is deployed as StatefulSet instead of Deployment by operator")
+				}
+			}
+		} else if kind == "Prometheus" && name == "prometheus" {
+			logrus.Warning("Prometheus service name will change from prometheus to px-prometheus after migration")
+			name = "px-prometheus"
+		} else if kind == "Service" && name == "prometheus" {
+			name = "px-prometheus"
+		} else if kind == "Service" && name == "autopilot" {
```

Comment:
Do we need to consider ccm here?

Reply:
ccm is part of telemetry. we dont support in daemonset spec however